### DISCUSSION TITLE
How to create a multi layer nested ListVector in C for an extension.

### DISCUSSION THREAD

**Discussion by HackettJP** [duckdb/duckdb#7125] [Q&A]
Hi i'm referencing the postgres_scanner extension trying to improve it for nested loops and also macaddr datatypes

Setup in Postgresql 

Data in Postgresql  " 2 rows , 1 column of text [] array" 
Row 1.    [a,b,c,d]
Row 2.    [e,f,g,h]

after looking at the vector list 

duck db will process this as a flat vector  [a,b,c,d,e,f,g,h] from start to end etc
with the list_entry.offset set to the row no. the list_entry=length.

Lines in postgres_scanner.c
 L-803    auto &list_entry = FlatVector::GetData<list_entry_t>(out_vec)[output_offset];
 L-804    auto child_offset = ListVector::GetListSize(out_vec);  // offset is the marker for a new row in postgresql.

Line 803 the output_offset is actually set to  the rowid within postgresql this is sort of confirmed in line 839 which eventually 
does this FlatVector::GetData<string_t>(out_vect)[child_offset+child_idx]  = StringVector::AddStringOrBlob(child_vec,data,length);
with child_idx as elements of the array....

I've been racking my brains on this last couple of days, and basically stuck... on how to implement a vector with parent then child and data vector  ? at least get two  -> three Dimensions  arrays into FlatVectors will prob solve majority of nested arrays from postgresql.

do you create another vector then reference that vector ? ReferenceEntry(Vector &vector, Vector &other), or is a DictionaryVector DictionaryVector::Child(out_vec);  then attach FlatVectors... 

now i can convert a valid postgresql {{a,b,c,d}}  or {{{{a,b,c,d}}}} -> to [a,b,c,d] from postgresql but works for analytics had in my data set were majority of GIS data is 2 dimension nested.

The next is to have postgresql {{a,b,c,d},{e,f,g,h},{........}} -> [[a,b,c,d],[e,f,g,h]........] in duckdb, 

this will also work for macaddr data types in duckdb from postgresql using a ByteA->BLOB is current way to display in duckdb but not do much, as VARCHAR will truncate the MAC address in duckdb.

Thks

James

**Comment by Tishj** [comment]
Looking at the `postgres_scanner.cpp` code I'm confused why this wouldn't already work for nested lists?
```c++
		for (idx_t child_idx = 0; child_idx < array_length; child_idx++) {
			// handle NULLs again (TODO: unify this with scan)
			auto ele_len = LoadEndIncrement<int32_t>(value_ptr);
			if (ele_len == -1) { // NULL
				FlatVector::Validity(child_vec).Set(child_offset + child_idx, false);
				continue;
			}

			ProcessValue(ListType::GetChildType(type), elem_info, atttypmod, 0, nullptr, value_ptr, ele_len, child_vec,
			             child_offset + child_idx);
			value_ptr += ele_len;
		}
```
It calls ProcessValue recursively for each of the entries of the list

**Reply by Tishj** [reply]
It sounds like you only need to make sure the type gets recursively resolved, seems like it becomes a 1 dimensional LIST always in the Bind.
```c++
static LogicalType DuckDBType2(PostgresTypeInfo *type_info, int atttypmod, PostgresTypeInfo *ele_info, PGconn *conn,
                               ClientContext &context) {
	auto &pgtypename = type_info->typname;

	// TODO better check, does the typtyp say something here?
	// postgres array types start with an _
	if (StringUtil::StartsWith(pgtypename, "_")) {
		return LogicalType::LIST(DuckDBType2(ele_info, atttypmod, nullptr, conn, context));
	}
```

**Comment by HackettJP** [comment]
hiya,

postgresql data is thus, and inside the LIST FUNCTION before its processed 

('{{ONE,TWO},{THREE,FOUR}}'); 

00 00 00 02 00 00 00 00  00 00 00 19 00 00 00 02  |  \000  \000  \000  \002  \000  \000  \000  \000  \000  \000  \000  \031  \000  \000  \000  \002 
00 00 00 01 00 00 00 02  00 00 00 01 00 00 00 03  |  \000  \000  \000  \001  \000  \000  \000  \002  \000  \000  \000  \001  \000  \000  \000  \003 
4F 4E 45 00 00 00 03 54  57 4F 00 00 00 05 54 48  |     O     N     E  \000  \000  \000  \003     T     W     O  \000  \000  \000  \005     T     H 
52 45 45 00 00 00 04 46  4F 55 52                 |     R     E     E  \000  \000  \000  \004     F     O     U     R 

The first 4 bytes describes Dimensions so this case is 2 mean a composite nested loop null bitmap flag or flag_two in the code, Iterate again to give you the matrix dimension size, then its no of elements in the array plus offset/lbounds, first 4 bytes decides  how many loops you need to do after the OID reference so duck db will not work you need to do a ndims loop unpack the array still 
you eventually have this sort of output from the header of VLA(postgres)
		               printf("\nNdims indicator {{.. <-no of (flagone)  :%u ", flag_one);    
				printf("\nvalue_oid 							:%u ", value_oid); 
				printf("\narray_length 	{x,x,x..}				:%u ", array_length);  
				printf("\narray_dim 							:%u ", array_dim);
				printf("\nntups is the number {},{},{} we have  :%d ", ntups); 

The detection of dimension is wrong it should just be set on flag_one only variable from ndims, inside of array_size.
if (flag_one > 1) {
			throw NotImplementedException("Only one-dimensional Postgres arrays currently supported, Detected %u dimensions", flag_one);
		}


  new bit of logic 
if (flag_one > 3) {
			throw NotImplementedException("Only max two-dimensional Postgres arrays currently supported, Detected %u dimensions", flag_one);
		}
//-- i've only done this for 2 dimension for easy of logic, you need to go arrays of flag_three, flag_four, created dynamically logic 
if (flag_one > 1 || flag_one < 3 ) {
			  for (int unpack = 1; unpack < flag_one; unpack++){
                       int *dims = (int *) malloc(flag_one * sizeof(int));
			int *lbound = (int *) malloc(flag_one * sizeof(int));
			           flag_three = (int)LoadEndIncrement<uint32_t>(value_ptr);  // array size
			          flag_four  = (int)LoadEndIncrement<uint32_t>(value_ptr);  // lbounds
			dims[unpack] 			= flag_three;
			lbound[unpack]			= flag_four;
			  printf("\n Dimension   :%d  matrix: %d , lbound %d  ntups (d) %d ", unpack,flag_three,flag_four,ntups);
			  ntups = ntups * flag_three;

			  }
                        //THEN 
                       array_length = LoadEndIncrement<int32_t>(value_ptr); 
		       array_dim = LoadEndIncrement<int32_t>(value_ptr);

                  //---- I either to a for loop or window function do be decided on performance as can be dims[i] for each unpack per layer , the this  case a simple 2 Dimension.
                       for (idx_t nest_idx = 0 ; nest_idx < ntups ; nest_idx++)
		   {
                  / //--- !!! MAGIC need to create parent 
                     for (idx_t child_idx = 0; child_idx < array_length; child_idx++) {
				//	printf("\n-==value_ptr (S) pos	 :%u ", value_ptr);
					// handle NULLs again (TODO: unify this with scan)
					auto ele_len = LoadEndIncrement<int32_t>(value_ptr);
...

 The recursive code just basically the [child_offset+child_idx] is the row of postgresql  data, i get that as its currently the row id, 
FlatVector::GetData<string_t>(out_vect)[child_offset+child_idx] = StringVector::AddStringOrBlob(child_vec,data,length);

I can see that the data is stored in duckdb or processed output as such:-

Row 1. {ONE,TWO,THREE,FOUR}
Row 2. {FIVE,SIX,SEVEN,EIGHT}

list entry is 
1 - offset 0, length 4  
2 - offset 4, length 4.
vector (VARCHAR)
ONE,TWO,THREE,FOUR,FIVE,SIX,SEVEN,EIGHT

I want to create this
 row 1. {{ONE,TWO,THREE,FOUR}, {ONE,EXTRA,LIKE,FOUR}}
 row 2. {{FIVE,SIX,SEVEN,EIGHT}.{DIFF,MORE,SOME,EIGHT}}

so it need to have a parent so my guess is has to go along the lines of with some magic. :)

for say a 2 Dimension nested array :-----
&parent_entry             = FlatVector::xxxxxx<list_entry_t>[row_id] 
     parent_entry           = ListVector::xxxx(out_vec)   
&child_entry                 = FlatVector::xxxxxx<list_entry_t>(parent_entry )[ntups]; // [ntups]  <-dims[1]
    child_entry                 =ListVector::xxxx(parent_entry )    
child_vector =
              ONE,TWO,THREE,FOUR,ONE,EXTRA,LIKE,FOUR,FIVE,SIX,SEVEN,EIGHT,DIFF,MORE,SOME,EIGHT

for say a 3 dimension array ;------
 row 1. {{{ONE,TWO},{THREE,FOUR}}, {{ONE,EXTRA},{LIKE,FOUR}}}
 row 2. {{{FIVE,SIX},{SEVEN,EIGHT}}.{{DIFF,MORE},{SOME,EIGHT}}}

&parent_entry = FlatVector::xxxxxx<list_entry_t>[row_id]
     parent_entry                      = ListVector::xxxx(out_vec)   
&child_entry = FlatVector::xxxxxx<list_entry_t>(parent_entry )[ntups]; // [ntups] <- dims[1]
    child_entry   =ListVector::xxxx(parent_entry )
&sub_child_entry = FlatVector::xxxxxx<list_entry_t>(childt_entry )[ntups];  // [ntups]   <- dims[2]
   sub_ child_entry   =ListVector::xxxx(child_entry )        
child_vector =
              ONE,TWO,THREE,FOUR,ONE,EXTRA,LIKE,FOUR,FIVE,SIX,SEVEN,EIGHT,DIFF,MORE,SOME,EIGHT


if you see what i mean logic thinking :) needs a slight rewrite, how to link a child to parent, i tried going down the path of this approach and was a no no.... 

int32_t list_entry = FlatVector::GetData<list_entry_t>(out_vec)[output_offset];

list_entry[0].
list_entry[0].......

James

**Reply by Tishj** [reply]
When constructing a Vector, you pass in a type, this is when the internals are initialized.
It sounds like you do know upfront how many dimensions there are (flag_one right?)

So you could construct a LIST type in a loop, and then create the Vector using the type, that should initialize all the buffers/auxiliary vectors properly. Handled by this code in `common/types/vector.cpp`
```c++
void Vector::Initialize(bool zero_data, idx_t capacity) {
	auxiliary.reset();
	validity.Reset();
	auto &type = GetType();
	auto internal_type = type.InternalType();
	if (internal_type == PhysicalType::STRUCT) {
		auto struct_buffer = make_uniq<VectorStructBuffer>(type, capacity);
		auxiliary = shared_ptr<VectorBuffer>(struct_buffer.release());
	} else if (internal_type == PhysicalType::LIST) {
		auto list_buffer = make_uniq<VectorListBuffer>(type, capacity);
		auxiliary = shared_ptr<VectorBuffer>(list_buffer.release());
	}
	auto type_size = GetTypeIdSize(internal_type);
	if (type_size > 0) {
		buffer = VectorBuffer::CreateStandardVector(type, capacity);
		data = buffer->GetData();
		if (zero_data) {
			memset(data, 0, capacity * type_size);
		}
	}
	if (capacity > STANDARD_VECTOR_SIZE) {
		validity.Resize(STANDARD_VECTOR_SIZE, capacity);
	}
}
```

So you likely want something like:
```c++
// not entirely sure where you can get the child type from
auto flag_one = LoadEndIncrement<uint32_t>(value_ptr);
type = LogicalType::LIST(child_type);
for (uint32_t i = 1; i < flag_one; i++) {
    type = LogicalType::LIST(type);
}
Vector list_vector(type, true, false, STANDARD_VECTOR_SIZE);
```
Though the bind code already initializes the vectors, so I'm not sure how this should work when (from what I understand you said) we only know the dimensions of the LIST once we're in ProcessValue

**Reply by HackettJP** [reply]
So delving deeper it in these two parts
![Screenshot from 2023-04-19 14-17-40](https://user-images.githubusercontent.com/71249639/232965896-9ef5773f-5516-4282-8eb7-5250d2981f60.png)

and the LIST properties

![Screenshot from 2023-04-19 14-18-33](https://user-images.githubusercontent.com/71249639/232966073-1a1105d6-2f3b-4594-a78b-0029b034f258.png)


so if we need to split the ::LIST function into posgresql VLA LIST header detection/decode and still have the main part for LIST if flag_one <1 if flag_one > 1  create a LIST_vector then call the ::LIST function 

LogicalType::LIST(DuckDBType2(ele_info, atttypmod, nullptr, conn, context));

seems a big rabbit hole , trying to understand ,thinking,approach, design executing in getting the list vector to work.
I'm used to C++ land, with BOOST, POCO STL's so much easier. 

J.

**Reply by Tishj** [reply]
You discovered that the dimensions of the list aren't part of the type info, but are instead part of the column info, right?

In the Bind we initialize the Vectors, but since we don't have all the information we need at that time, I think we have to reinitialize the Vector received in ProcessValue.

Once the list vector is initialized (again, I think this only requires the snippet I send before), then you should just be able to populate it

**Reply by HackettJP** [reply]
Correct, 

Postgresql will only store in the DB typeinfos that is an array is present, there is a  column in postgresql that is meant to show the number of dimensions at creation of table, column, however that is not enforced (on data input or output from postgresql DB) it was was never implemented "More than one it an nested array of unknown size") ,  so you have to, decode the Variable Header to get the size ,width, depth, etc.

INSERT INTO public.baddoubledoubletext VALUES ('{ONE}');
INSERT INTO public.baddoubledoubletext VALUES ('{F_ONE,F_TWO}');
INSERT INTO public.baddoubledoubletext VALUES ('{NULL}');
INSERT INTO public.baddoubledoubletext VALUES ('{{NULL,DO_TWO}}'); -- NOTE NDIMS (2) but single ntups.

with the decoding you can get away with row 4 in postgresl to be ingested into duckdb minus the extra [] etc..

<pre>
Ndims indicator {{{..}}} no of : 1 
Null Bitmap Flag Set           : false 
value_len                      : 27 
value_oid                      : 25 
internal_position              : 7 
array_length  {x,x,x..}        : 1 
array_dim                      : 1 
child_offset                   : 0 
ntups  no of {},{},{} we in{..}: 1 
00 00 00 03 4F 4E 45                              |  \000  \000  \000  \003     O     N     E 

-==target (Child_VEC) child_offset.output_offset    :0 , (child_offset.offset : 0 + child_idx  : 0) 
-==length of data field (value_len)ele_len          :3 , datafield {x,x,x..}   array.length     : 1

-==SetListSize (OUT_VEC) elements.offset            :0 , datafield {x,x,x..} list_entry.length : 1 

Ndims indicator {{{..}}} no of : 1 
Null Bitmap Flag Set           : false 
value_len                      : 38 
value_oid                      : 25 
internal_position              : 18 
array_length  {x,x,x..}        : 2 
array_dim                      : 1 
child_offset                   : 1 
ntups  no of {},{},{} we in{..}: 1 
00 00 00 05 46 5F 4F 4E  45 00 00 00 05 46 5F 54  |  \000  \000  \000  \005     F     _     O     N     E  \000  \000  \000  \005     F     _     T 
57 4F                                             |     W     O 

-==target (Child_VEC) child_offset.output_offset    :1 , (child_offset.offset : 1 + child_idx  : 0) 
-==length of data field (value_len)ele_len          :5 , datafield {x,x,x..}   array.length     : 2

-==target (Child_VEC) child_offset.output_offset    :2 , (child_offset.offset : 1 + child_idx  : 1) 
-==length of data field (value_len)ele_len          :5 , datafield {x,x,x..}   array.length     : 2

-==SetListSize (OUT_VEC) elements.offset            :1 , datafield {x,x,x..} list_entry.length : 2 

Ndims indicator {{{..}}} no of : 1 
Null Bitmap Flag Set           : true 
value_len                      : 24 
value_oid                      : 25 
internal_position              : 4 
array_length  {x,x,x..}        : 1 
array_dim                      : 1 
child_offset                   : 3 
ntups  no of {},{},{} we in{..}: 1 
FF FF FF FF                                       |  \377  \377  \377  \377 

-==SetListSize (OUT_VEC) elements.offset            :3 , datafield {x,x,x..} list_entry.length : 1 

Ndims indicator {{{..}}} no of : 2 
Null Bitmap Flag Set           : true 
value_len                      : 42 
value_oid                      : 25 
internal_position              : 14 
array_length  {x,x,x..}        : 2 
array_dim                      : 1 
child_offset                   : 4 
ntups  no of {},{},{} we in{..}: 1 
FF FF FF FF 00 00 00 06  44 4F 5F 54 57 4F        |  \377  \377  \377  \377  \000  \000  \000  \006     D     O     _     T     W     O 

-==target (Child_VEC) child_offset.output_offset    :5 , (child_offset.offset : 4 + child_idx  : 1) 
-==length of data field (value_len)ele_len          :6 , datafield {x,x,x..}   array.length     : 2
44 4F 5F 54 57 4F                                 |     D     O     _     T     W     O 

-==SetListSize (OUT_VEC) elements.offset            :4 , datafield {x,x,x..} list_entry.length : 2 
┌────────────────┐
│    column1     │
│   varchar[]    │
├────────────────┤
│ [ONE]          │
│ [F_ONE, F_TWO] │
│ [NULL]         │
│ [NULL, DO_TWO] │
└────────────────┘
</pre>


        ntups = 1;  // Use to times the with ntups to get the matrix size
	    if (flag_one > 1 ){
			// Decode composite nested array in postgresql
			int *dims = (int *) malloc(flag_one * sizeof(int));
			int *lbound = (int *) malloc(flag_one * sizeof(int));

			for (idx_t unpack = 1; unpack < flag_one; unpack++){
						flag_three = (int)LoadEndIncrement<uint32_t>(value_ptr); 
						flag_four  = (int)LoadEndIncrement<uint32_t>(value_ptr);
                         // required if NDIMS > 3 for each  {{{layer }}} will have its own associated, ntups {{},{},{}} size
						dims[unpack] 			= flag_three;   
 						lbound[unpack]			= flag_four;    

                        internal_position = internal_position - 8;
                        // Total ntups calculation total arrays 
						ntups = ntups * flag_three;  
				}

				auto child_offset   = ListVector::GetListSize(out_vec);
				auto array_length   = LoadEndIncrement<uint32_t>(value_ptr);
				auto array_dim      = LoadEndIncrement<uint32_t>(value_ptr);
                internal_position   = internal_position-8;
                printf("\nNdims indicator {{{..}}} no of : %u ", flag_one);
                printf("\nNull Bitmap Flag Set           : %s ", flag_two ? "true" : "false");
                printf("\nvalue_len                      : %u ", value_len);    
                printf("\nvalue_oid                      : %u ", value_oid);
                printf("\ninternal_position              : %u ", internal_position);
                printf("\narray_length  {x,x,x..}        : %u ", array_length);  
                printf("\narray_dim                      : %u ", array_dim);
                printf("\nchild_offset                   : %u ", child_offset);
                printf("\nntups  no of {},{},{} we in{..}: %d \n", ntups);

                if (flag_one > 2) {
					throw NotImplementedException("more than 2-dimensional {{...}} Postgres arrays currently supported, Detected %u dimensions", flag_one);
				}			
                // only added in to allow  {{ONE,TWO}} to be accepted atm translates to {ONE,TWO,THREE} output
				if ( ntups > 1) {  
					throw NotImplementedException("more than 2-dimension matrix with 1-LIST {{ONE,TWO,THREE}} currently supported, Detected %u dimensions with ntups %u ", flag_one,flag_three);
				}

                // you times ntups with array_length to get get reserve size
                // {{x.x},{y.y}}  be described as ntups (2) array_length (2)
                // {{x.x}}  be described as ntups (1) array_length (2)
            
				ListVector::Reserve(out_vec, child_offset+(array_length*ntups));

               // CREATE A LOOP 
                // type = LogicalType::LIST(child_type);
                // Vector list_vector(type, true, false, STANDARD_VECTOR_SIZE);
                // example {x.x.},{y.y}  be described as ntups (2) array_length (2)
                // list_vector.entry=0 , length =2
                // list_vector.entry=2 , length =2,
                // list_vector.entry= ntups ; list_vector.length =array_length;
               // END LOOP 
         
				auto &child_vec = ListVector::GetEntry(out_vec);
				auto &child_mask = FlatVector::Validity(child_vec);

                HexAsciiDump((unsigned char *)value_ptr,internal_position); 
				for (idx_t child_idx = 0; child_idx < array_length; child_idx++) {
					
					// handle NULLs again (TODO: unify this with scan)
					auto ele_len = LoadEndIncrement<int32_t>(value_ptr);
					
					if (ele_len == -1) { // NULL
					
						child_mask.Set(child_offset + child_idx, false);
						continue;
						}	

                printf("\n-==target (Child_VEC) child_offset.output_offset    :%u , (child_offset.offset : %u + child_idx  : %u) ",child_offset+child_idx,child_offset,child_idx);
                printf("\n-==length of data field (value_len)ele_len          :%u , datafield {x,x,x..}   array.length     : %u\n",ele_len,array_length); 	 				
							HexAsciiDump((unsigned char *)value_ptr,ele_len);
                            ProcessValue(ListType::GetChildType(type), elem_info, atttypmod, 0, nullptr, value_ptr, ele_len, child_vec,child_offset + child_idx);
							value_ptr += ele_len;
					}
                ListVector::SetListSize(out_vec, child_offset + array_length);
                list_entry.offset = child_offset;
                list_entry.length = array_length;
                printf("\n-==SetListSize (OUT_VEC) elements.offset            :%u , datafield {x,x,x..} list_entry.length : %u \n", child_offset, array_length);
                break;

		}



type = LogicalType::LIST(child_type);
Vector list_vector(type, true, false, STANDARD_VECTOR_SIZE);

I'm really not sure on what the child_type be linked to ,  as you said can be re-initialised ???.... 

i really appreciated the helping out.

J.

**Reply by Tishj** [reply]
Please send a link to your branch, it's really hard to follow without seeing the code

**Reply by HackettJP** [reply]
Tishj

https://github.com/HackettJP/postgres_scanner_array/tree/HackettJP-Ndimension_WIP

you have to disable address option(ENABLE_SANITIZER "Enable address sanitizer." TRUE) in the  CmakeLists if you need to compile
FOR SOME TEST CASES

CREATE TABLE public.texttable
(
column1 text []
);


INSERT INTO public.texttable VALUES ('{ONE}');
INSERT INTO public.texttable VALUES ('{F_ONE,F_TWO}');
INSERT INTO public.texttable VALUES ('{NULL}');
INSERT INTO public.texttable VALUES ('{{NULL,DO_TWO}}'); -- NOTE NDIMS (2) but single ntups 

---- LAST ENTRY will exception out,  but will be able to see the decode params
INSERT INTO public.texttable VALUES ('{{NULL,DO_TWO},{NULL,DO_THREE}}'); -- NOTE NDIMS (2) and ntups is (2) ntups 

James

**Reply by Tishj** [reply]
So I was thinking of changing it to this:
```c++
			auto list_type = LogicalType::LIST(ListType::GetChildType(out_vec.GetType()));
			for (idx_t unpack = 1; unpack < flag_one; unpack++) {
				flag_three = (int)LoadEndIncrement<uint32_t>(value_ptr);
				flag_four = (int)LoadEndIncrement<uint32_t>(value_ptr);
				// required if NDIMS > 3 for each  {{{layer }}} will have its own associated, ntups {{},{},{}} size
				dims[unpack] = flag_three;
				lbound[unpack] = flag_four;
				internal_position = internal_position - 8;
				// Total ntups calculation total arrays
				list_type = LogicalType::LIST(list_type);
				ntups = ntups * flag_three;
			}

			out_vec = Vector(list_type, true, false, STANDARD_VECTOR_SIZE);
```
But the problem is that `Vector` doesn't have a copy constructor, so we can't just change the type after it's created.
Even if that were possible, I think the execution would fail, because the whole point of the Bind phase is that we know which types are returned.

**Reply by Tishj** [reply]
I think the bind code needs to be changed so we can properly detect the amount of dimensions of the List column there

**Reply by HackettJP** [reply]
T,
could you just point at the bind code and then i can read up :) thks
J.

**Comment by HackettJP** [comment]
thks

J.
